src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs

```diff
@@ -196,15 +196,15 @@ internal static class DynamicMethodBuilder<TDelegate>
 
             if (methodInfo.ReturnType.IsValueType && returnType == typeof(object))
             {
-                dynamicMethod.Box(methodInfo.ReturnType);
+                il.Emit(OpCodes.Box, methodInfo.ReturnType);
             }
             else if (methodInfo.ReturnType != returnType)
```

Comment:
If `returnType` is a value type, the `Castclass` is invalid.

Reply:
You're right. If I use your suggestion to check if the returnType is a value type instead of just strictly equal to `typeof(object)`, I have the following code, but I don't really know what we should be doing in this additional case. Throw an exception since the method was incorrectly generated by us?

```csharp
            if (methodInfo.ReturnType.IsValueType && !returnType.IsValueType) // Changed from returnType == typeof(object)
            {
                il.Emit(OpCodes.Box, methodInfo.ReturnType);
            }
            else if (methodInfo.ReturnType.IsValueType && methodInfo.ReturnType != returnType)
            {
                // Figure out what to do
            }
            else if (!methodInfo.ReturnType.IsValueType && methodInfo.ReturnType != returnType)
            {
                il.Emit(OpCodes.Castclass, returnType);
            }
```

Reply:
Yeah, I will expect an exception on those cases. (Internally we can just return without converting between two value types. That way is how Unsafe.As works but is dangerous to do that without being explicit)

Reply:
Another case I hadn't considered: If the target method is an object but the outer delegate is a value type. Should we attempt an unbox and if it's incompatible then that operation will throw an InvalidClassException, like so:

```csharp
            if (methodInfo.ReturnType.IsValueType && !returnType.IsValueType)
            {
                il.Emit(OpCodes.Box, methodInfo.ReturnType);
            }
            else if (methodInfo.ReturnType.IsValueType && returnType.IsValueType && methodInfo.ReturnType != returnType)
            {
                throw new ArgumentException($"Cannot convert the target method's return type {methodInfo.ReturnType.FullName} (valuetype) to the delegate method's return type {returnType.FullName} (valuetype)");
            }
            else if (!methodInfo.ReturnType.IsValueType && returnType.IsValueType)
            {
                il.Emit(OpCodes.Unbox, returnType);
            }
            else if (!methodInfo.ReturnType.IsValueType && !returnType.IsValueType && methodInfo.ReturnType != returnType)
            {
                il.Emit(OpCodes.Castclass, returnType);
            }
```

Reply:
```
if (!methodInfo.ReturnType.IsValueType && returnType.IsValueType)
```
IMO, I prefer to throw an exception and not handle it as a best effort. As developer we have control of the `returnType`. Is an edge case a scenario where the target method returns an object and we want a value type. If the target method returns a class, we should change the returnType to that class or an object in case we don't have the class at compile time.

```
 if (methodInfo.ReturnType.IsValueType && returnType.IsValueType && methodInfo.ReturnType != returnType)
```
Now I'm thinking in this scenario where the target method returns an struct that we don't have at compile time. Actually what we can do is to create our own struct with the same layout at the target and just return without any conversion. I don't know if we ever gonna have that scenario, but at least there is a way to handle it in the future if we want to.

Reply:
The latter scenario with different value types is really interesting. If we have that scenario in the future, I think we can pick that up then

Reply:
👍